WIP: Add BorrowingSequence above Sequence

natecook1000 opened the pull request:
Adds a new `BorrowingSequence` and makes `Sequence` extend the new protocol. 

#### Issues still to tackle:
- unsure how to handle availability for `Sequence` protocol when its declaration would require the availability-constrained `BorrowingSequence` – availability checking currently off
- conditionally-conforming `Sequence` types (like `Range<Int>`) don't implicitly conform to `BorrowingSequence`
- commented out both a chunk of `SwiftOnone.swift` specialization code and the bit of the compiler that verifies the existence of those prespecialized symbols

natecook1000 committed:
Initial work on BorrowingSequence above Sequence

So far this runs into the issue that conditionally-conforming Sequence
types (like Range<Int>) don't implicitly conform to BorrowingSequence.

natecook1000 committed:
Work around conditional Sequence conformance

benrimmington commented:
[Library Evolution in Swift](<https://www.swift.org/blog/library-evolution/#examples-of-non-resilient-changes>):

> Changes to a protocol’s list of refined protocols are not allowed.

natecook1000 committed:
Require Copyable for same-type BorrowingIterator

Returning self from this default impl of makeBorrowingInterator() is
consuming, so it requires a copyable type (like Span).

natecook1000 committed:
BorrowingSequence conformances for Array/InlineArray

natecook1000 committed:
Work around SwiftOnone specialization issues

This disables the SwiftOnone code that triggers errors like:

Cannot pre-specialize $sSMsSKRzrlE14_insertionSort6within2byySny5IndexSlQzG_Sb7ElementSTQz_AHtKXEtKF

and also disables the diagnostics when those prespecialized symbols
can no longer be found.

natecook1000 committed:
Enable tests for BorrowingSequence

natecook1000 committed:
Modify Sequence extension with BorrowingSequence requirements

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Remove availability on BorrowingSequence

lorentey commented on stdlib/public/core/BorrowingSequence.swift:
```diff
@@ -0,0 +1,207 @@
+  /// in bulk, by directly iterating over its piecewise contiguous pieces of
+  /// storage:
+  ///
+  ///     var it = items.startBorrowIteration()
```
```suggestion
  ///     var it = items.makeBorrowingIterator()
```

natecook1000 committed:
Relax Span BorrowingSequence constraint

Also includes conformances for other stdlib types
